[HUDI-499] Allow update partition path with GLOBAL_BLOOM

xushiyan opened the pull request:
## What is the purpose of the pull request

Allow records to be inserted into their new partition paths and delete the records in the old partition 
paths. 

## Brief change log

- Add a configuration `hoodie.index.bloom.should.update.partition.path` with default `false`
- When the config set to `true`, perform the partition update for incoming records of which partition paths were changed

## Verify this pull request

This pull request is already covered by existing tests, such as *(please describe tests)*.

## Committer checklist

 - [x] Has a corresponding JIRA in PR title & commit
 
 - [x] Commit message is descriptive of the change
 
 - [x] CI is green

 - [x] Necessary doc changes in #1190

vinothchandar commented:
@nsivabalan could you please review this?

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -77,6 +77,9 @@
   public static final String BLOOM_INDEX_INPUT_STORAGE_LEVEL = "hoodie.bloom.index.input.storage.level";
   public static final String DEFAULT_BLOOM_INDEX_INPUT_STORAGE_LEVEL = "MEMORY_AND_DISK_SER";
 
+  public static final String BLOOM_INDEX_SHOULD_UPDATE_PARTITION_PATH = "hoodie.bloom.index.should.update.partition.path";
```
should we include "global" in the name. Bcoz, this is applicable only incase of global. In case of regular, the new record will be inserted as a new entry to new partition path.

nsivabalan replied:
Also, can we add some java docs since the property is not very intuitive.

xushiyan replied:
Renamed and added docs.

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -431,6 +431,10 @@ public StorageLevel getBloomIndexInputStorageLevel() {
     return StorageLevel.fromString(props.getProperty(HoodieIndexConfig.BLOOM_INDEX_INPUT_STORAGE_LEVEL));
   }
 
+  public boolean getBloomIndexShouldUpdatePartitionPath() {
```
would appreciate if you can add java docs

xushiyan replied:
@nsivabalan I think for a simple a getter that fetch the property, the javadoc is accessible from the property itself. I've added the docs there. So can i skip it here?

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java:
```diff
@@ -114,14 +117,23 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
         // Record key matched to file
-        return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
+        if (config.getBloomIndexShouldUpdatePartitionPath()) {
+          HoodieRecord<T> emptyRecord = new HoodieRecord(recordLocationHoodieKeyPair.get()._2,
```
I guess we need to have a if else case here. Let me try to explain my understanding. correct me if its wrong. 
The scenario we are trying to tackle is, insertion for a record happened in PartitionPath1, while update is sent to PartitionPath2. With the newly added cofig value set to true, we want the update operation to insert in PartitionPath2 and also delete in PartitionPath1. 

So, in this case, after the left outer join, here are the values for the record
record._1 -> incoming record. recordKey1, PartitionPath2
record._2 -> Tuple2<HoodieRecordLocation, HoodieKey> after index look up. should refer to recordKey1, PartitionPath1. 

`
      if (recordLocationHoodieKeyPair.isPresent()) {

        if(recordLocationHoodieKeyPair.get()._2.getPartitionPath().equals(hoodieRecord.getPartitionPath())){
          return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
        } else {
          if(config value is true) {
            // need to add two records. one delete in old partition path and an insert into new one.
            HoodieRecord<T> emptyRecord = new HoodieRecord(recordLocationHoodieKeyPair.get()._2,
                new EmptyHoodieRecordPayload());
            HoodieRecord<T> taggedRecord = getTaggedRecord(hoodieRecord, Option.empty());
            return Arrays.asList(emptyRecord, taggedRecord).iterator();
            return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
          } else{
            return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
          }
        }

        // Record key matched to file
       // return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
      } else {
        return getTaggedRecord(hoodieRecord, Option.empty());
      }
`

xushiyan replied:
Exactly @nsivabalan . I understood the logic but missed the part of checking path update.

nsivabalan commented:
@vinothchandar : sure. I will take care.

xushiyan commented:
@nsivabalan Thank you for the thorough review! I'll try to address these in the next few days.

xushiyan commented:
@nsivabalan @vinothchandar I'll squash the commits if the new changes look good. Thank you!

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -77,6 +77,17 @@
+  /**
+   * Only applies if index type is GLOBAL_BLOOM.
+   * <p>
+   * When set to true, an update including the partition path of a record that already exists will result in inserting
```
minor: "when set to true, an update to a record with diff partition compared to its existence, will insert the record to new partition and delete from old partition. When set to false, record will be updated to old partition".

xushiyan replied:
> can we also add this to the configurations md file , to keep the docs upto date?

The docs update PR is in #1190 . I'll update it once this is merged.

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java:
```diff
@@ -114,14 +117,24 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
-        return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
+        if (config.getGlobalBloomIndexShouldUpdatePartitionPath()
+            && !recordLocationHoodieKeyPair.get()._2.getPartitionPath().equals(hoodieRecord.getPartitionPath())) {
+          HoodieRecord<T> emptyRecord = new HoodieRecord(recordLocationHoodieKeyPair.get()._2,
```
can we add a comment here. for eg "If partition path differs from incoming record to those in index, and if config for partition path update is set to true, 
a. insert to new partition from incoming record 
b. delete existing record from existing partition from index".

xushiyan replied:
Ok I'll just explain the purposes of those 2 `HoodieRecord`s, which I think it'd be clear enough.

vinothchandar replied:
+1

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java:
```diff
@@ -114,14 +117,24 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
+          HoodieRecord<T> taggedRecord = getTaggedRecord(hoodieRecord, Option.empty());
+          return Arrays.asList(emptyRecord, taggedRecord).iterator();
+        } else {
+          return Collections.singletonList(
```
can we add a comment here mentioning which branch this else part serves. For eg, "Either if partition path matches from incoming record to those in index, or if config value for partition path update is set to false, choose the partition path from index and update the record, ignoring the partition from incoming record".

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java:
```diff
@@ -114,14 +117,24 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
       if (recordLocationHoodieKeyPair.isPresent()) {
         // Record key matched to file
-        return getTaggedRecord(new HoodieRecord<>(recordLocationHoodieKeyPair.get()._2, hoodieRecord.getData()), Option.ofNullable(recordLocationHoodieKeyPair.get()._1));
+        if (config.getGlobalBloomIndexShouldUpdatePartitionPath()
```
minor: Would be good if we first check partition path differs and then check config value. Bcoz, the config value doesn't matter if partition path matches. So, for readability purposes would be good to swap the condition checks.

vinothchandar replied:
I can actually make the argument otherwise :) .. the first check is easier to read and I only need to both with the second one, if first passes.. we can leave this to @xushiyan  to make the call.

xushiyan replied:
2nd thought on this: we should treat a configuration as the dominating factor and the partition equality check as a supporting factor. Doing it the other way will diminish the purpose of a configuration. Hence I would keep that order as is. Thanks for bringing thoughts to it!

nsivabalan commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -77,6 +77,17 @@
+   * <p>
+   * When set to false, the original record will only be updated in the old partition.
+   */
+  public static final String GLOBAL_BLOOM_INDEX_SHOULD_UPDATE_PARTITION_PATH = "hoodie.global.bloom.index.should.update.partition.path";
```
@vinothchandar : does the config param looks good. I am okay with this. just wanted your opinion.

vinothchandar replied:
Since this applies to the bloom index, can we slightly rename to `hoodie.bloom.index.update.partition.path` and also shorten it a bit

xushiyan replied:
> Since this applies to the bloom index, can we slightly rename to `hoodie.bloom.index.update.partition.path` and also shorten it a bit

I like the brevity. As the docs specify as "only apply to GLOBAL_BLOOM", it is unlikely to confuse users.

xushiyan commented:
@nsivabalan @vinothchandar I've addressed the suggestions. Please check the last 2 commits. Shall I squash it once approved? Or will you use the GitHub feature to squash it?

nsivabalan commented:
LGTM

xushiyan commented:
@vinothchandar Squashing done. Thanks.

vinothchandar commented:
@xushiyan I think there is a checkstyle error. can you please take a look

xushiyan commented:
> @xushiyan I think there is a checkstyle error. can you please take a look

Fixed!

vinothchandar commented:
Thanks and Congrats! :) landed!
